core/src/docs/rfcs/2602_object_versioning.md

```diff
@@ -0,0 +1,78 @@
+Proposal Name: object versioning
+Start Date: 2023-07-06
+RFC PR: https://github.com/apache/incubator-opendal/pull/2602
+Tracking Issue: (leave this empty)
+
+# Summary
+
+This proposal describes the object versioning (or object version control) feature of OpenDAL.
+
+# Motivation
+
+Object versioning is a common feature in many storage services. 
+
+With object versioning, users can:
+
+- Track the history of an object.
+- Implement optimistic concurrency control.
+- Implement a simple backup system.
+
+# Guide-level explanation
+
+When object versioning is enabled, each object will have a history of versions. Each version will have a unique version ID, which is a string that is unique for each version of an object.
+
+The version ID is not a timestamp. It is not guaranteed to be sequential.
+
+When object versioning is enabled, the following operations will be supported:
+
+- `stat`: Get the metadata of an object with specific version ID.
+- `read`: Read a specific version of an object.
+- `delete`: Delete a specific version of an object.
+- `list`: List all versions of an object.
+
+Those operations with object version are different from the normal operations:
+
+- `write`: when writing an object, it will always create a new version of the object than overwrite the old version. But here it is imperceptible to the user. Because the version id is generated by the service itself, it cannot be specified by the user and user cannot override the historical version. 
+- `stat`: when getting the metadata of an object, it will always get the metadata of the latest version of the object if no version ID is specified.
+- `read`: when reading an object, it will always read the latest version of the object if no version ID is specified.
+- `delete`: when deleting an object, it will always delete the latest version of the object if no version ID is specified. And users will not be able to read this object unless they specify the version ID not to be deleted.
+- `list_versions`: this `list_versions` means list all versions of an object. It's an operation on a single object, and it is different from the normal `list` which means list all objects.
+
+```rust
+// read with version ID
+let content = op.read_with("path/to/file").with_version("version_id").await?;
+```
+
+# Reference-level explanation
```

Comment:
In this section, we need to explanation what's the API semantice and how will we implement it. For example

- What's field we will add in `OpRead`, `OpStat`.
- What changes we will do for services.